### PR TITLE
Add python packages: pyensembl and its dependencies, gffutils

### PR DESCRIPTION
**Pull request by moritzschaefer**
##### Motivation for this change

Replaces PR https://github.com/NixOS/nixpkgs/pull/87551

Both pyenseml and gffutils are commonly used python packages (> 100 stars on GitHub), so I added them to nix packages. 

###### Things done

- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`) (only pyensembl has a supplemental binary)
- [x] Ensured that relevant documentation is up to date
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [-] Mostly Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md). I designed the derivations 1:1 to existing ones.

### CODE REVIEW

File: pkgs/development/python-modules/datacache/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  };
+
+  checkInputs = [ ];
+  propagatedBuildInputs = [ pandas appdirs progressbar33 requests typechecks nose mock ];
+
+  checkPhase = ''
+      ${python.interpreter} -c 'import datacache'
+  '';
+
+  doCheck = true;
```
Comment (jonringer) [inline pkgs/development/python-modules/datacache/default.nix:20]:
```suggestion
  propagatedBuildInputs = [ pandas appdirs progressbar33 requests typechecks nose mock ];

  pythonImportsCheck = [ "datacache" ];
```

File: pkgs/development/python-modules/datacache/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  meta = with stdenv.lib; {
+    homepage = "https://github.com/openvax/datacache";
+    description = "Helpers for transparently downloading datasets";
+    license = licenses.asl20;
```
Comment (jonringer) [inline pkgs/development/python-modules/datacache/default.nix:25]: please add yourself as a maintainer

File: pkgs/development/python-modules/datacache/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  version = "1.1.5";
+  pname = "datacache";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "b2ca31b2b9d3803a49645ab4f5b30fdd0820e833a81a6952b4ec3a68c8ee24a7";
+  };
+
+  checkInputs = [ ];
+  propagatedBuildInputs = [ pandas appdirs progressbar33 requests typechecks nose mock ];
```
Comment (jonringer) [inline pkgs/development/python-modules/datacache/default.nix:14]: not needed at runtime, should go in checkInputs if anything
```suggestion
  propagatedBuildInputs = [ pandas appdirs progressbar33 requests typechecks mock ];
```

File: pkgs/development/python-modules/gffutils/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    inherit pname version;
+    sha256 = "a8fc39006d7aa353147238160640e2210b168f7849cb99896be3fc9441e351cb";
+  };
+
+  checkInputs = [ nose wget ];
+  propagatedBuildInputs = [ pyfaidx six argh argcomplete simplejson ];
+
+  checkPhase = ''
+    # sh gffutils/test/data/download-large-annotation-files.sh
+    # nosetests
```
Comment (jonringer) [inline pkgs/development/python-modules/gffutils/default.nix:17]: why no test suite?
```suggestion
    # nosetests
```

File: pkgs/development/python-modules/gtfparse/default.nix [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    sha256 = "2f27aa2b87eb43d613edabf27f9c11147dc595c8683b440ac1d88e9acdb85873";
+  };
+
+  checkInputs = [ nose six ];
+  propagatedBuildInputs = [ numpy pandas ];
+
+  checkPhase = ''
+    # PYTHONPATH='test' nosetests # fails because six is not found
+    ${python.interpreter} -c 'import gtfparse'
+  '';
```
Comment (jonringer) [inline pkgs/development/python-modules/gtfparse/default.nix:19]: use pythonImportsCheck

File: pkgs/development/python-modules/memoized-property/default.nix [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    sha256 = "4be4d0209944b9b9b678dae9d7e312249fe2e6fb8bdc9bdaa1da4de324f0fcf5";
+  };
+
+  checkInputs = [ ];
+  propagatedBuildInputs = [ ];
+
+  checkPhase = ''
+      ${python.interpreter} -c 'import memoized_property'
+  '';
+
```
Comment (jonringer) [inline pkgs/development/python-modules/memoized-property/default.nix:18]: use pythonImportsCheck

File: pkgs/development/python-modules/pybedtools/default.nix [hunk 7]
```diff
@@ -0,0 +1,29 @@
+{ bash, stdenv, buildPythonPackage, fetchPypi, numpy, pandas, pysam, six,
+  pytest, numpydoc, psutil, pyyaml, sphinx, zlib, bedtools, python
+}:
```
Comment (jonringer) [inline pkgs/development/python-modules/pybedtools/default.nix:3]:
```suggestion
}:

```

File: pkgs/development/python-modules/pybedtools/default.nix [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    inherit pname version;
+    sha256 = "c035e078617f94720eb627e20c91f2377a7bd9158a137872a6ac88f800898593";
+  };
+
+  checkInputs = [ pytest numpydoc psutil pyyaml sphinx ];
+  propagatedBuildInputs = [ numpy pandas pysam six zlib bash bedtools ];
+
+  checkPhase = ''
+    # pytest -v --doctest-modules
+    # ${python.interpreter} -c 'import pybedtools'  # test and import do not work in checkPhase, because the built pyx file cannot be included
```
Comment (jonringer) [inline pkgs/development/python-modules/pybedtools/default.nix:18]: pythonImportsCheck

File: pkgs/development/python-modules/pyensembl/default.nix [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "13dd05aba296e4acadb14de5a974e6f73834452851a36b9237917ae85b3e060f";
+  };
+
+  checkInputs = [ ];
+  propagatedBuildInputs = [ numpy pandas datacache six memoized-property gtfparse serializable tinytimer serializable ];
+
+  checkPhase = ''
+    ${python.interpreter} -c 'import pyensembl'  # fails only in build environment because pyensembl creates a file in root directory
```
Comment (jonringer) [inline pkgs/development/python-modules/pyensembl/default.nix:17]: pythonImportsCheck

File: pkgs/development/python-modules/scikit-plot/default.nix [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+buildPythonPackage rec {
+  version = "0.3.7";
+  pname = "scikit-plot";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "2c7948817fd2dc06879cfe3c1fdde56a8e71fa5ac626ffbe79f043650baa6242";
+  };
+
+  buildInputs = [  ];
```
Comment (jonringer) [inline pkgs/development/python-modules/scikit-plot/default.nix:21]:
```suggestion
```

File: pkgs/development/python-modules/scikit-plot/default.nix [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  checkInputs = [ nose ];
+  propagatedBuildInputs = [ matplotlib scikitlearn scipy joblib ];
+
+  doCheck = true;
+
```
Comment (jonringer) [inline pkgs/development/python-modules/scikit-plot/default.nix:26]: default for `buildPythonPackage`
```suggestion;
```

File: pkgs/development/python-modules/serializable/default.nix [hunk 12]
```diff
@@ -0,0 +1,27 @@
+{ stdenv, buildPythonPackage, fetchPypi,
+typechecks, six, simplejson, nose}:
```
Comment (jonringer) [inline pkgs/development/python-modules/serializable/default.nix:2]:
```suggestion
typechecks, six, simplejson, nose}:

```

File: pkgs/development/python-modules/serializable/default.nix [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  checkInputs = [ nose ];
+  propagatedBuildInputs = [ typechecks six simplejson ];
+
+  checkPhase = ''
+    nosetests
+  '';
+
+  # Tests require extra dependencies
+  doCheck = true;
+
```
Comment (jonringer) [inline pkgs/development/python-modules/serializable/default.nix:21]: default
```suggestion
```

File: pkgs/development/python-modules/tinytimer/default.nix [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+  checkInputs = [ ];
+  propagatedBuildInputs = [ ];
+
+  checkPhase = ''
+    ${python.interpreter} -c 'import tinytimer'
+  '';
+
+  # Tests require extra dependencies
+  doCheck = true;
+
```
Comment (jonringer) [inline pkgs/development/python-modules/tinytimer/default.nix:21]:
```suggestion
  pythonImportsCheck = [ "tinytimer" ];
```

File: pkgs/top-level/python-packages.nix [hunk 15]
```diff
@@ -7405,6 +7405,25 @@ in {
...
+
+  tinytimer = callPackage ../development/python-modules/tinytimer { };
+
+  pyensembl = callPackage ../development/python-modules/pyensembl { };
+
+  gffutils = callPackage ../development/python-modules/gffutils { };
+
+  scikit-plot = callPackage ../development/python-modules/scikit-plot { };
+
+  pybedtools = callPackage ../development/python-modules/pybedtools { };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:7426]: please attempt to sort these

**Review by jonringer — COMMENTED** [review]
please add yourself as a maintainer for all packages that you're adding.

please attempt to sort the package entries in `python-packages`

File: pkgs/development/python-modules/gffutils/default.nix (hunk 4 above)
Comment (moritzschaefer) [reply pkgs/development/python-modules/gffutils/default.nix:17]: The tests fail inside the nix environment

**Comment by moritzschaefer** [comment]
Thank you for your comments @jonringer! I have simplified all the files and added myself as maintainer. Please let me know if something still requires improvements.

File: pkgs/top-level/python-packages.nix (hunk 15 above)
Comment (jonringer) [reply pkgs/top-level/python-packages.nix:7426]: it has been evolving over time... some people grouped similar packages, some people just dropped them anywhere. At the very least, random listings should be avoided

File: pkgs/development/python-modules/gffutils/default.nix (hunk 4 above)
Comment (jonringer) [reply pkgs/development/python-modules/gffutils/default.nix]: general consensus is to just ignore test cases that fail due to sandbox environment. But some test failures may be indications that the package is in an actual broken state.

tests are also very useful for verifying that dependency bumps don't cause regressions.

**Comment by moritzschaefer** [comment]
@jonringer After having applied all the requested changes, I wanted to da a final rebase on master so I did

> git checkout python-packages
> git pull --rebase official master  # official points to github.com/NIXOS/nixpgs.git

This resulted in a merge commit and now the PR is blown up with 370 commits. Is this an issue and if yes, would you mind telling me what went wrong?

**Comment by jonringer** [comment]
with --rebase, there shouldn't be a merge commit.

And it looks like you pulled in master. Assuming the merge is the main culprit, you could do `git reset HEAD^ --hard` but, I'm not sure exactly what you've done.

If all else failes, you could cherry pick the commits into a new branch and just push that

**Comment by moritzschaefer** [comment]
Indeed trashing the mere commit seemed to have fixed it. So no cherry-picking-in-new-branch required. 
Would you approve to merge like this?

**Comment by jonringer** [comment]
if you still want to get this into nixpkgs, you can break this up into smaller PRs, just adding a package at a time. It will be a lot less time intensive than having to cover all of it.

**Comment by moritzschaefer** [comment]
Dear @jonringer,

thanks for the response. I'll transition to adding these packages into my overlays and close this PR as there are definitely more important PRs (from me and others) for the moment.

Still, I am kind of missing some kind of strategy for pypi packages. In a conversation on the nixos-discourse, I got the notion that it is not desired to add the majority of pypi packages to nixpkgs. However it is really cumbersome to build a derivation for each single python package, when you're working in a non-mainstream field. 

Is there some kind of vision for this issue?

**Comment by jonringer** [comment]
In general the preference is to have python packages available which enable an application, and less so to "just have the package". 



> However it is really cumbersome to build a derivation for each single python package, when you're working in a non-mainstream field.

You can use nix with venv to bring in missing python packages. I got a video about it https://www.youtube.com/watch?v=jXd-hkP4xnU , and some docs here: https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#how-to-consume-python-modules-using-pip-in-a-virtual-environment-like-i-am-used-to-on-other-operating-systems

If you still want a pure nix way to do it, I would look at `mach-nix`https://github.com/DavHau/mach-nix

**Pull request closed without merging**